File: 백준/Silver/9012. 괄호/괄호.c
#define _CRT_SECURE_NO_WARNINGS
#include <stdio.h>
char stack[50];
int top = 0;

void push(char c) {
	stack[top++] = c;
}

char pop() {
	return stack[--top];
}

int isEmpty() {
	return top == 0;
}

int isVPS(char* str) {
	int result = 1;
	int i;

	for (i = 0; str[i]; i++) {
		if (str[i] == '(')
			push(str);
		else {
			if (isEmpty()) {
				result = 0;
				break;
			}
			else pop();
		}
	}
	if (!isEmpty()) result = 0;
	while (!isEmpty()) {
		pop();
	}
	return result;
}
int main()
{
	int testn;
	char str[51];
	int i;

	scanf("%d", &testn);
	for (i = 0; i < testn; i++) {
		scanf("%s", str);
		if (isVPS(str)) printf("YES\n");
		else printf("NO\n");		
	}
	return 0;
}

File: 백준/Silver/9012. 괄호/괄호.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.Stack;

public class Main {
    public static void main(String[] args) throws IOException {

        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringBuilder sb = new StringBuilder();

        int N = Integer.parseInt(br.readLine());

        String[] input = new String[51];

        for(int i = 0; i < N; i++){
            sb.append(isVPS(br.readLine())).append('\n');
        }

        System.out.println(sb);
    }
    static String isVPS(String str){
        Stack<Character> stack = new Stack<>();
        for(int i = 0; i < str.length(); i++) {
            char c = str.charAt(i);

            if(c == '(')
                stack.push(c);
            else if(stack.isEmpty())
                return "NO";
            else
                stack.pop();
        }

        if(stack.isEmpty())
            return "YES";
        else
            return "NO";
    }
}
